src/Integration/Connection/ConnectionWorkflow.cs

```diff
@@ -112,34 +114,80 @@ private void OnProjectsChanged(ConnectionInformation connection, ProjectInformat
             {
+                AbortWithMessage(notifications, controller, cancellationToken);
                 return;
             }
 
+            if (connection.Organization == null &&
+                this.host.SonarQubeService.HasOrganizationsSupport(connection, cancellationToken))
+            {
+                notifications.ProgressChanged(Strings.ConnectionStepRetrievingOrganizations);
+                OrganizationInformation[] organizations;
+                if (!this.host.SonarQubeService.TryGetOrganizations(connection, cancellationToken, out organizations))
+                {
+                    AbortWithMessage(notifications, controller, cancellationToken);
+                    return;
+                }
+
+                if (organizations.Length <= 1)
+                {
+                    connection.Organization = null;
+                }
+                else
+                {
+                    connection.Organization = AskUserToSelectOrganizationOnUIThread(organizations);
+                }
+            }
+
+            if (!this.AreSolutionProjectsAndServerPluginsCompatible(controller, cancellationToken, connection, notifications))
+            {
+                return; // Message is already displayed by the method
+            }
+
             this.ConnectedServer = connection;
 
+            notifications.ProgressChanged(Strings.ConnectionStepRetrievingProjects);
             ProjectInformation[] projects;
             if (!this.host.SonarQubeService.TryGetProjects(connection, cancellationToken, out projects))
             {
-                notifications.ProgressChanged(cancellationToken.IsCancellationRequested ? Strings.ConnectionResultCancellation : Strings.ConnectionResultFailure);
-                this.host.ActiveSection?.UserNotifications?.ShowNotificationError(Strings.ConnectionFailed, NotificationIds.FailedToConnectId, this.parentCommand);
-
-                AbortWorkflow(controller, cancellationToken);
+                AbortWithMessage(notifications, controller, cancellationToken);
                 return;
             }
 
             this.OnProjectsChanged(connection, projects);
             notifications.ProgressChanged(Strings.ConnectionResultSuccess);
         }
 
+        private void AbortWithMessage(IProgressStepExecutionEvents notifications, IProgressController controller, CancellationToken cancellationToken)
+        {
+            notifications.ProgressChanged(cancellationToken.IsCancellationRequested ? Strings.ConnectionResultCancellation : Strings.ConnectionResultFailure);
+            this.host.ActiveSection?.UserNotifications?.ShowNotificationError(Strings.ConnectionFailed, NotificationIds.FailedToConnectId, this.parentCommand);
+
+            AbortWorkflow(controller, cancellationToken);
+        }
+
+        private OrganizationInformation AskUserToSelectOrganizationOnUIThread(IEnumerable<OrganizationInformation> organizations)
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Rename method 'AskUserToSelectOrganizationOnUIThread' to match camel case naming rules, consider using 'AskUserToSelectOrganizationOnUiThread'. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS100)